src/Operations.jl

```diff
@@ -1524,8 +1524,10 @@ function add(ctx::Context, pkgs::Vector{PackageSpec}, new_git=Set{UUID}();
             compat_names = String[]
             for pkg in pkgs
                 haskey(ctx.env.project.compat, pkg.name) && continue
-                pkgversion = Base.thispatch(ctx.env.manifest[pkg.uuid].version)
-                set_compat(ctx.env.project, pkg.name, string(pkgversion))
+                v = ctx.env.manifest[pkg.uuid].version
+                v === nothing && continue
```

Comment:
I think we'd expect the package version to be populated here by then by e.g. `update_manifest!`.
The scenario you have isn't that the manifest ends up without a version, right?

Reply:
Yeah, that makes sense - I was worried that might be the case.

The MWE is just `Pkg.add(Pkg.PackageSpec(name="SuiteSparse", uuid="4607b0f0-06f3-5cda-b6b1-a6196a1729e9"))`

If I apply this workaround to get past the error message it looks like we do end up with a missing version in the Manifest:
```julia
[[deps.SuiteSparse]]
deps = ["Libdl", "LinearAlgebra", "Serialization", "SparseArrays"]
uuid = "4607b0f0-06f3-5cda-b6b1-a6196a1729e9"

[[deps.SuiteSparse_jll]]
deps = ["Artifacts", "Libdl", "libblastrampoline_jll"]
uuid = "bea87d4a-7f5b-5778-9afe-8cc45184846c"
version = "7.7.0+0"
```

Reply:
Looks like the same thing happens on 1.10 too

Reply:
@KristofferC I think we made it that stdlibs get versions recorded in the manifest, so this seems like a bug?

Reply:
I feel we have a real issue in Pkg as described in https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ denoted as "shotgun parsing".

> Shotgun parsing is a programming antipattern whereby parsing and input-validating code is mixed with and spread across processing code—throwing a cloud of checks at the input, and hoping, without any systematic justification, that one or another would catch all the “bad” cases.

We have this `PackageSpec` object that:

- Users add fields to from the constructor.
- During the progress of a Pkg operation we discover various things and update those fields.

But it is hard to know what state the PackageSpec is in at some arbitrary point in the code. Is it valid or do more things have to be "discovered" on the way. In https://github.com/JuliaLang/Pkg.jl/pull/3783 this was particularly annoying because this was yet another way where new information was discovered and that surprised the existing code etc.

Reply:
This at least gets us back to the (admittedly a bit incorrect) behavior that we have on 1.10

Can we merge this and open up an issue for the deeper problem here?